drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1026 @@
+
+	uint8_t id[2];
+	uint8_t support_conti_read;
+};
+
+/**
+ * struct spi_nand_data - Structure for defining the SPI NAND access
+ * @sem: The semaphore to access to the flash
+ */
+struct spi_nand_data {
+	struct k_sem sem;
+
+	/* Number of bytes per page */
+	uint16_t page_size;
+
+	/* Number of oob bytes per page */
+	uint8_t oob_size;
+
+	/* Number of pages per block */
+	uint16_t page_num;
+
+	/* Number of bytes per block */
+	uint32_t block_size;
+
+	/* Number of blocks per chip */
+	uint16_t block_num;
+
+	/* Size of flash, in bytes */
+	uint32_t flash_size;
+
+	uint8_t page_shift;
+
+	uint8_t block_shift;
+
+	bool continuous_read;
+
+	bool read_recovery;
+
+	uint8_t ecc_bits;
+
+	uint8_t ecc_bytes;
+	uint8_t ecc_steps;
+	uint8_t ecc_layout_pos;
+	uint32_t ecc_size;
+	uint8_t *ecc_calc;
+	uint8_t *ecc_code;
+	uint8_t *page_buf;
+
+	struct nand_bch_control {
+		struct bch_code      *bch;
+		unsigned int         *errloc;
+		unsigned char        *eccmask;
+	} nbc;
+};
+
+static const struct flash_parameters flash_nand_parameters = {
+	.write_block_size = SPI_NAND_SUB_PAGE_SIZE,
+	.erase_value = 0xff,
+};
+
+void bch_calculate_ecc(const struct device *dev,
+		unsigned char *buf, unsigned char *code)
+{
+	struct spi_nand_data *data = dev->data;
+
+	memset(code, 0, data->ecc_bytes);
+
+	bch_encode(data->nbc.bch, buf, (unsigned int *)code);
+}
+
+void bch_correct_data(const struct device *dev, 
+		unsigned char *buf, unsigned char *read_ecc,
+		unsigned char *calc_ecc)
+{
+	struct spi_nand_data *data = dev->data;
+
+	bch_decode(data->nbc.bch, buf, (unsigned int *)calc_ecc);
+}
+
+/*
+ * @brief Send an SPI command
+ *
+ * @param dev Device struct
+ * @param opcode The command to send
+ * @param access flags that determine how the command is constructed.
+ *        See NAND_ACCESS_*.
+ * @param addr The address to send
+ * @param data The buffer to store or read the value
+ * @param length The size of the buffer
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nand_access(const struct device *const dev,
+			  uint8_t opcode, unsigned int access,
+			  off_t addr, void *data, size_t length)
+{
+	const struct spi_nand_config *const driver_cfg = dev->config;
+	bool is_addressed = (access & NAND_ACCESS_ADDRESSED) != 0U;
+	bool is_write = (access & NAND_ACCESS_WRITE) != 0U;
+	uint8_t buf[5] = { 0 };
+	uint8_t address_len = 0;
+	struct spi_buf spi_buf[2] = {
+		{
+			.buf = buf,
+			.len = 1,
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+
+	buf[0] = opcode;
+	if (is_addressed) {
+		union {
+			uint32_t u32;
+			uint8_t u8[4];
+		} addr32 = {
+			.u32 = sys_cpu_to_be32(addr),
+		};
+
+		if (access & NAND_ACCESS_32BIT_ADDR) {
+			address_len = 4;
+		} else if (access & NAND_ACCESS_24BIT_ADDR){
+			address_len = 3;
+		} else if (access & NAND_ACCESS_16BIT_ADDR){
+			address_len = 2;
+		} else if (access & NAND_ACCESS_8BIT_ADDR){
+			address_len = 1;
+		}
+		memcpy(&buf[1], &addr32.u8[3], address_len);
```

Comment:
memcpy(&buf[1], &addr32.u8[3], address_len); incorrectly optimizes out multiple calls to memcpy.

previously:
```
if (access & NAND_ACCESS_32BIT_ADDR)
{
	memcpy(&buf[1], &addr32.u8[0], 4);
	spi_buf[0].len += 4;
}
else if (access & NAND_ACCESS_24BIT_ADDR)
{
	memcpy(&buf[1], &addr32.u8[1], 3);
	spi_buf[0].len += 3;
}
else if (access & NAND_ACCESS_16BIT_ADDR)
{
	memcpy(&buf[1], &addr32.u8[2], 2);
	spi_buf[0].len += 2;
}
else if (access & NAND_ACCESS_8BIT_ADDR)
{
	memcpy(&buf[1], &addr32.u8[3], 1);
	spi_buf[0].len += 1;
}
```
therefore, 
memcpy(&buf[1], &addr32.u8[3], address_len); 
should be changed to:
memcpy(&buf[1], &addr32.u8[4 - address_len], address_len);

to match previous operation

Reply:
resolved